### PR TITLE
Check for MPI-3 by looking for MPI_NO_OP

### PR DESCRIPTION
**Pull request by devreal**
Check for MPI-3 by looking for MPI_NO_OP

See also #393

### CODE REVIEW

File: CMakeExt/MPI.cmake [hunk 1]
```diff
@@ -35,3 +37,18 @@ if (MPI_INCLUDE_PATH AND MPI_LIBRARY)
...
 endif (MPI_INCLUDE_PATH AND MPI_LIBRARY)
+
+
+# check for MPI-3
+
+set(CMAKE_REQUIRED_INCLUDES ${MPI_INCLUDE_PATH})
+CHECK_SYMBOL_EXISTS(
+  MPI_NO_OP
+  mpi.h
+  HAVE_MPI_NOOP
```
Comment (bertwesarg) [inline CMakeExt/MPI.cmake]: Why not `HAVE_MPI_NO_OP`?

File: CMakeExt/MPI.cmake [hunk 2]
```diff
@@ -35,3 +37,18 @@ if (MPI_INCLUDE_PATH AND MPI_LIBRARY)
 else (MPI_INCLUDE_PATH AND MPI_LIBRARY)
   set(MPI_FOUND FALSE CACHE BOOL "Did not find the MPI library")
 endif (MPI_INCLUDE_PATH AND MPI_LIBRARY)
+
+
+# check for MPI-3
+
+set(CMAKE_REQUIRED_INCLUDES ${MPI_INCLUDE_PATH})
```
Comment (bertwesarg) [inline CMakeExt/MPI.cmake]: I'm just curious, what is the scope of this `set`? Is it bound to this `.cmake` file? Or should we reset it after the `CHECK_SYMBOL_EXISTS` call. Neither the documention nor the `CheckSymbolExists.cmake` file gives me a clue about this.

File: CMakeExt/MPI.cmake [hunk 3]
```diff
@@ -35,3 +37,18 @@ if (MPI_INCLUDE_PATH AND MPI_LIBRARY)
...
+CHECK_SYMBOL_EXISTS(
+  MPI_NO_OP
+  mpi.h
+  HAVE_MPI_NOOP
+)
+
+if (NOT HAVE_MPI_NOOP)
+  message(FATAL_ERROR "Detected MPI library does not support MPI-3.")
```
Comment (bertwesarg) [inline CMakeExt/MPI.cmake]: While I find this rather harsh to the user, it is the RightThing™ to abort here.

**Review by bertwesarg — APPROVED** [review]
Tested with Open MPI 1.5 and 1.6.

File: CMakeExt/MPI.cmake (hunk 2 above)
Comment (bertwesarg) [reply CMakeExt/MPI.cmake]: Just found `CMAKE_PUSH_CHECK_STATE` and `CMAKE_POP_CHECK_STATE` in `CMakePushCheckState.cmake`. I propose to surround the set and check with these two calls.

Comment (fmoessbauer) [reply CMakeExt/MPI.cmake]: Looks interesting and is supported even in cmake 2.8

File: CMakeExt/MPI.cmake [hunk 4]
```diff
@@ -35,3 +37,21 @@ if (MPI_INCLUDE_PATH AND MPI_LIBRARY)
 else (MPI_INCLUDE_PATH AND MPI_LIBRARY)
   set(MPI_FOUND FALSE CACHE BOOL "Did not find the MPI library")
 endif (MPI_INCLUDE_PATH AND MPI_LIBRARY)
+
+
+# check for MPI-3
+
+# save current state
+cmake_push_check_state(RESET)
```
Comment (bertwesarg) [inline CMakeExt/MPI.cmake]: Why the RESET?

File: CMakeExt/MPI.cmake [hunk 5]
```diff
@@ -35,3 +37,21 @@ if (MPI_INCLUDE_PATH AND MPI_LIBRARY)
...
+CHECK_SYMBOL_EXISTS(
+  MPI_NO_OP
+  mpi.h
+  HAVE_MPI_NO_OP
+)
+
+if (NOT HAVE_MPI_NO_OP)
+  message(FATAL_ERROR "Detected MPI library does not support MPI-3.")
+endif()
+cmake_pop_check_state()
```
Comment (bertwesarg) [inline CMakeExt/MPI.cmake]: I would move this up directly after the check, line 52.

**Review by bertwesarg — APPROVED** [review]

**Pull request merged**